### PR TITLE
#314 Performances of models are not comparable across experiments or reproductions

### LINKED ISSUE #314: Performances of models are not comparable across experiments or reproductions
## 🐛 Bug description

We cannot easily or automatically compare the performances of models across experiments or reproductions.

This is especially critical when re-running our DVC pipelines.

## To reproduce

1. `dvc repro -f`
2. `dvc metrics diff`

Nothing is displayed.

## Expected behavior

There should be a table with the performances before and after, including the delta between the two.

Files holding the metrics **should be tracked with Git and not with DVC** as of now.
This is the best practice advertised by DVC (see reference below). 

The fix is to add `cache: false` to the `metrics` files in `dvc.yaml`.

## Additional context

Reference: https://dvc.org/doc/start/metrics-parameters-plots#get-started-metrics-parameters-and-plots.

## Environment

 - Blue Brain Search version: 0.1.1.dev35.
 - OS: macOS.
 - How you installed Blue Brain Search: source.
 - Installation command you used: pip install.
 - Python version: Python 3.8.5
 - Any other relevant information: issue spotted while working on #274.

### PR DESCRIPTION
**Pull request by EmilieDel**
Fixes #314.

## Description

> `dvc metrics diff` should display table with the performances before and after, including the delta between the two. Files holding the metrics should be tracked with Git and not with DVC as of now.

During this PR:
- All `dvc run` commands outputting `metrics` file are re-run. All metrics file are now specified as `metrics files` instead of `output files` (`dvc run -M metric_file` instead of `dvc_run -o metric_file`). 
From those commands, `dvc.yaml` changed:
  - `cache: false` is appearing after each metric file. [expected changes]
  - `cmd` command line is slightly changing and `deps` order is slightly different too [not expected but do not change anything]
- `.gitignore` files were updated and all `metrics files` uploaded to git. 

## How to test?
First:
```
dvc repro
dvc metrics diff
```
No table should be displayed out of this.

Secondly: (`BBS_REVISION=v0.1.2`)
1. Modify `dropout` inside `config.cfg` from `0.2` to `0.1` for example. (line 86)
2. `git diff` is going to show
```
@@ -83,7 +83,7 @@ train_corpus = "corpora.train"
 dev_corpus = "corpora.dev"
 seed = ${system.seed}
 gpu_allocator = ${system.gpu_allocator}
-dropout = 0.2
+dropout = 0.1
 accumulate_gradient = 1
 patience = 1600
 max_epochs = 0
```
3. Run same command lines as for the first experiment: (This might takes some time)
```
dvc repro
dvc metrics diff
```
This time, the last command should give a summary table:
```
Path                                     Metric            Old      New      Change
../../metrics/ner/drug.json              entity_f1-score   0.62667  0.55208  -0.07458
../../metrics/ner/drug.json              entity_precision  0.63514  0.4569   -0.17824
../../metrics/ner/drug.json              entity_recall     0.61842  0.69737  0.07895
../../metrics/ner/drug.json              token_f1-score    0.62893  0.56716  -0.06177
../../metrics/ner/drug.json              token_precision   0.65789  0.48305  -0.17484
../../metrics/ner/drug.json              token_recall      0.60241  0.68675  0.08434
../../metrics/ner/organ.json             entity_f1-score   0.54545  0.54475  -0.00071
../../metrics/ner/organ.json             entity_precision  0.42105  0.42683  0.00578
../../metrics/ner/organ.json             entity_recall     0.77419  0.75269  -0.02151
../../metrics/ner/organ.json             token_f1-score    0.58632  0.60751  0.02119
../../metrics/ner/organ.json             token_precision   0.46154  0.49171  0.03017
../../metrics/ner/organ.json             token_recall      0.80357  0.79464  -0.00893
../../metrics/ner/cell_type.json         entity_precision  0.61017  0.63303  0.02286
../../metrics/ner/cell_type.json         entity_recall     0.73469  0.70408  -0.03061
../../metrics/ner/cell_type.json         token_f1-score    0.76883  0.75493  -0.0139
../../metrics/ner/cell_type.json         token_precision   0.6758   0.70899  0.0332
../../metrics/ner/cell_type.json         token_recall      0.89157  0.80723  -0.08434
../../metrics/ner/organism.json          entity_f1-score   0.56917  0.55217  -0.017
../../metrics/ner/organism.json          entity_precision  0.48322  0.50397  0.02075
../../metrics/ner/organism.json          entity_recall     0.69231  0.61058  -0.08173
../../metrics/ner/organism.json          token_f1-score    0.67445  0.63507  -0.03938
../../metrics/ner/organism.json          token_precision   0.62097  0.62813  0.00716
../../metrics/ner/organism.json          token_recall      0.73802  0.64217  -0.09585
../../metrics/ner/disease.json           entity_f1-score   0.72262  0.71802  -0.00459
../../metrics/ner/disease.json           entity_precision  0.65974  0.66757  0.00783
../../metrics/ner/disease.json           entity_recall     0.79874  0.77673  -0.02201
../../metrics/ner/disease.json           token_f1-score    0.77891  0.77074  -0.00817
../../metrics/ner/disease.json           token_precision   0.72126  0.74483  0.02357
../../metrics/ner/disease.json           token_recall      0.84658  0.79852  -0.04806
...
```
(The results is longer). 

## Checklist

- [x] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [x] Unit tests added.
  (if needed)
- [ ] Documentation and `whatsnew.rst` updated.
  (if needed)
- [x] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [x] Type annotations added.
  (if a function is added or modified)
- [x] All CI tests pass.

### CODE REVIEW

**Review by pafonta — APPROVED** [review]
Thank you very much @EmilieDel for fixing #314!

That's something I missed so much on #248!

By the way, one could also use `dvc params diff` to see the change in parameters :)

Now we can do the following and that's awesome!

```
/pipelines/ner# dvc params diff
Path         Param                                         Old    New 
params.yaml  train.model5.prodigy_train_kwargs.eval_split  0.1    0.3

pipelines/ner# dvc metrics diff
Path                            Metric            Old      New      Change
../../metrics/ner/pathway.json  entity_f1-score   0.58039  0.53125  -0.04914
../../metrics/ner/pathway.json  entity_precision  0.68519  0.62385  -0.06133
../../metrics/ner/pathway.json  entity_recall     0.5034   0.46259  -0.04082
../../metrics/ner/pathway.json  token_f1-score    0.58586  0.55138  -0.03448
../../metrics/ner/pathway.json  token_precision   0.71166  0.66265  -0.04901
../../metrics/ner/pathway.json  token_recall      0.49785  0.4721   -0.02575
```

**Comment by FrancescoCasalegno** [comment]
I am not sure about one thing: the `dvc repro` is referred to which version of `bluesearch`?
I am asking because `v0.1.1` got released recently and `v0.1.2` may be released soon (see #274).

Also: what is the relation between this PR and #274? Should this PR (#323) get merged first, wouldn't this imply the need of `git add` the new scores on the side of #274?

**Comment by EmilieDel** [comment]
> I am not sure about one thing: the dvc repro is referred to which version of bluesearch?
> I am asking because v0.1.1 got released recently and v0.1.2 may be released soon (see #274).

Currently, `dvc repro` is referring to `v0.1.1`. I was thinking about waiting for #274 to be merged (as it is already huge changes) and updating this PR once the new version is released. This PR might change a lot but it is going to keep the same spirit. If you think it is useful, we can put this issue as blocked and wait for #274.

**Comment by FrancescoCasalegno** [comment]
> I was thinking about waiting for #274 to be merged (as it is already huge changes)

Sure it makes sense.

> This PR might change a lot but it is going to keep the same spirit. If you think it is useful, we can put this issue as blocked and wait for #274.

Sounds like a good idea! Please move into Blocked then.

**Comment by pafonta** [comment]
Hello @EmilieDel,

Thank you very much for having rebased on top of #248 after it was merged!

I still approve the changes.

File: data_and_models/pipelines/sentence_embedding/dvc.yaml
Comment (FrancescoCasalegno) [inline data_and_models/pipelines/sentence_embedding/dvc.yaml:28]: `.png` outputs are pretty quick to recreate, and at the same time each of these output images takes ~0.2 MB (on the other hand, `.csv` are just ~4 KB each).

So what about declaring the `.png` outputs as `cache: false` as well?

**Review by FrancescoCasalegno — CHANGES_REQUESTED** [review]
This PR definitely enables a _really_ useful feature, thank you!

I have left a few comments, and I think one point in the `whatsnew.rst` needs to be fixed before this PR can be merged.

File: data_and_models/pipelines/sentence_embedding/dvc.yaml
Comment (Stannislav) [reply data_and_models/pipelines/sentence_embedding/dvc.yaml:28]: Would this mean we would track these PNG-files in our GitHub repository instead of the DVC remote?

**Review by Stannislav — APPROVED** [review]
I haven't tested the functionality, but the code changes look good to me.

Would you recomment testing it for real?

**Review by jankrepl — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**
